Jackson + Records requires reflection configuration

### Expected Behavior

Should work without reflection configuration

### Actual Behaviour

Fails with:

```
Caused by: java.lang.IllegalArgumentException: Failed to access RecordComponents of type `java17test.Book`
        at com.fasterxml.jackson.databind.jdk14.JDK14Util$RecordAccessor.recordComponents(JDK14Util.java:124)
        at com.fasterxml.jackson.databind.jdk14.JDK14Util$RecordAccessor.getRecordFieldNames(JDK14Util.java:78)
        at com.fasterxml.jackson.databind.jdk14.JDK14Util.getRecordFieldNames(JDK14Util.java:27)
        at com.fasterxml.jackson.databind.introspect.DefaultAccessorNamingStrategy$RecordNaming.<init>(DefaultAccessorNamingStrategy.java:531)
        at com.fasterxml.jackson.databind.introspect.DefaultAccessorNamingStrategy$Provider.forRecord(DefaultAccessorNamingStrategy.java:448)
        at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.collectProperties(BasicClassIntrospector.java:185)
        at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.forSerialization(BasicClassIntrospector.java:86)
        at com.fasterxml.jackson.databind.introspect.BasicClassIntrospector.forSerialization(BasicClassIntrospector.java:11)
        at com.fasterxml.jackson.databind.SerializationConfig.introspect(SerializationConfig.java:862)
        at com.fasterxml.jackson.databind.ser.BeanSerializerFactory.createSerializer(BeanSerializerFactory.java:139)
        at com.fasterxml.jackson.databind.SerializerProvider._createUntypedSerializer(SerializerProvider.java:1473)
        at com.fasterxml.jackson.databind.SerializerProvider._createAndCacheUntypedSerializer(SerializerProvider.java:1421)
        ... 35 common frames omitted
```

### Steps To Reproduce

1. Create a JDK 17 Record and annotate it with `@Introspected`
2. Serialize to JSON

### Environment Information

_No response_

### Example Application

_No response_

### Version

3.0.3

For future reference, there is another exception when deserializing:

```
Caused by: java.lang.IllegalArgumentException: Failed to access RecordComponents of type `core6277.Rec`
        at com.fasterxml.jackson.databind.jdk14.JDK14Util$RecordAccessor.recordComponents(JDK14Util.java:124)
        at com.fasterxml.jackson.databind.jdk14.JDK14Util$RecordAccessor.getRecordFields(JDK14Util.java:94)
        at com.fasterxml.jackson.databind.jdk14.JDK14Util$CreatorLocator.<init>(JDK14Util.java:155)
        at com.fasterxml.jackson.databind.jdk14.JDK14Util.findRecordConstructor(JDK14Util.java:32)
        at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory._constructDefaultValueInstantiator(BasicDeserializerFactory.java:287)
        at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory.findValueInstantiator(BasicDeserializerFactory.java:223)
        at com.fasterxml.jackson.databind.deser.BeanDeserializerFactory.buildBeanDeserializer(BeanDeserializerFactory.java:261)
        ... 12 more
```

--- a/jackson-databind/src/main/java/io/micronaut/jackson/ObjectMapperFactory.java
+++ b/jackson-databind/src/main/java/io/micronaut/jackson/ObjectMapperFactory.java
@@ -26,7 +26,10 @@ import com.fasterxml.jackson.databind.JsonSerializer;
 import com.fasterxml.jackson.databind.KeyDeserializer;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.fasterxml.jackson.databind.PropertyNamingStrategy;
+import com.fasterxml.jackson.databind.cfg.DeserializerFactoryConfig;
+import com.fasterxml.jackson.databind.deser.BeanDeserializerFactory;
 import com.fasterxml.jackson.databind.deser.BeanDeserializerModifier;
+import com.fasterxml.jackson.databind.deser.DefaultDeserializationContext;
 import com.fasterxml.jackson.databind.deser.std.StringDeserializer;
 import com.fasterxml.jackson.databind.module.SimpleModule;
 import com.fasterxml.jackson.databind.ser.BeanSerializerModifier;
@@ -111,7 +114,7 @@ public class ObjectMapperFactory {
     public ObjectMapper objectMapper(@Nullable JacksonConfiguration jacksonConfiguration,
                                      @Nullable JsonFactory jsonFactory) {
 
-        ObjectMapper objectMapper = jsonFactory != null ? new ObjectMapper(jsonFactory) : new ObjectMapper();
+        ObjectMapper objectMapper = new ObjectMapper(jsonFactory, null, new DefaultDeserializationContext.Impl(new ResilientBeanDeserializerFactory(new DeserializerFactoryConfig())));
 
         final boolean hasConfiguration = jacksonConfiguration != null;
         if (!hasConfiguration || jacksonConfiguration.isModuleScan()) {

--- a/jackson-databind/src/main/java/io/micronaut/jackson/ResilientBeanDeserializerFactory.java
+++ b/jackson-databind/src/main/java/io/micronaut/jackson/ResilientBeanDeserializerFactory.java
@@ -1,3 +1,18 @@
+/*
+ * Copyright 2017-2022 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
 package io.micronaut.jackson;
 
 import com.fasterxml.jackson.databind.BeanDescription;

--- a/jackson-databind/src/main/java/io/micronaut/jackson/ResilientBeanDeserializerFactory.java
+++ b/jackson-databind/src/main/java/io/micronaut/jackson/ResilientBeanDeserializerFactory.java
@@ -1,0 +1,44 @@
+package io.micronaut.jackson;
+
+import com.fasterxml.jackson.databind.BeanDescription;
+import com.fasterxml.jackson.databind.DeserializationConfig;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.cfg.DeserializerFactoryConfig;
+import com.fasterxml.jackson.databind.deser.BeanDeserializerFactory;
+import com.fasterxml.jackson.databind.deser.DeserializerFactory;
+import com.fasterxml.jackson.databind.deser.ValueInstantiator;
+import com.fasterxml.jackson.databind.deser.impl.CreatorCollector;
+
+/**
+ * This class hooks {@link BeanDeserializerFactory} to fix an exception in {@link #_constructDefaultValueInstantiator}
+ * that happens when jackson attempts to access record member info in a native-image without reflective access for the
+ * record class.
+ *
+ * @author yawkat
+ * @since 3.3.0
+ */
+final class ResilientBeanDeserializerFactory extends BeanDeserializerFactory {
+    public ResilientBeanDeserializerFactory(DeserializerFactoryConfig config) {
+        super(config);
+    }
+
+    @Override
+    public DeserializerFactory withConfig(DeserializerFactoryConfig config) {
+        return new ResilientBeanDeserializerFactory(config);
+    }
+
+    @Override
+    protected ValueInstantiator _constructDefaultValueInstantiator(DeserializationContext ctxt, BeanDescription beanDesc) throws JsonMappingException {
+        try {
+            return super._constructDefaultValueInstantiator(ctxt, beanDesc);
+        } catch (IllegalArgumentException e) {
+            if (e.getMessage().startsWith("Failed to access RecordComponents of type ")) {
+                final DeserializationConfig config = ctxt.getConfig();
+                return new CreatorCollectionState(ctxt, beanDesc, config.getDefaultVisibilityChecker(beanDesc.getBeanClass(), beanDesc.getClassInfo()),
+                        new CreatorCollector(beanDesc, config), _findCreatorsFromProperties(ctxt, beanDesc)).creators.constructValueInstantiator(ctxt);
+            }
+            throw e;
+        }
+    }
+}

--- a/jackson-databind/src/main/java/io/micronaut/jackson/modules/BeanIntrospectionModule.java
+++ b/jackson-databind/src/main/java/io/micronaut/jackson/modules/BeanIntrospectionModule.java
@@ -41,7 +41,6 @@ import com.fasterxml.jackson.databind.deser.impl.MethodProperty;
 import com.fasterxml.jackson.databind.deser.std.StdValueInstantiator;
 import com.fasterxml.jackson.databind.introspect.AccessorNamingStrategy;
 import com.fasterxml.jackson.databind.introspect.AnnotatedClass;
-import com.fasterxml.jackson.databind.introspect.AnnotatedField;
 import com.fasterxml.jackson.databind.introspect.AnnotatedMember;
 import com.fasterxml.jackson.databind.introspect.AnnotatedMethod;
 import com.fasterxml.jackson.databind.introspect.AnnotationCollector;

--- a/jackson-databind/src/main/java/io/micronaut/jackson/modules/BeanIntrospectionModule.java
+++ b/jackson-databind/src/main/java/io/micronaut/jackson/modules/BeanIntrospectionModule.java
@@ -24,6 +24,7 @@ import com.fasterxml.jackson.annotation.JsonUnwrapped;
 import com.fasterxml.jackson.annotation.OptBoolean;
 import com.fasterxml.jackson.core.JsonGenerator;
 import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.ObjectCodec;
 import com.fasterxml.jackson.core.SerializableString;
 import com.fasterxml.jackson.core.io.SerializedString;
 import com.fasterxml.jackson.databind.*;
@@ -38,8 +39,13 @@ import com.fasterxml.jackson.databind.deser.SettableBeanProperty;
 import com.fasterxml.jackson.databind.deser.ValueInstantiator;
 import com.fasterxml.jackson.databind.deser.impl.MethodProperty;
 import com.fasterxml.jackson.databind.deser.std.StdValueInstantiator;
+import com.fasterxml.jackson.databind.introspect.AccessorNamingStrategy;
+import com.fasterxml.jackson.databind.introspect.AnnotatedClass;
+import com.fasterxml.jackson.databind.introspect.AnnotatedField;
 import com.fasterxml.jackson.databind.introspect.AnnotatedMember;
+import com.fasterxml.jackson.databind.introspect.AnnotatedMethod;
 import com.fasterxml.jackson.databind.introspect.AnnotationCollector;
+import com.fasterxml.jackson.databind.introspect.DefaultAccessorNamingStrategy;
 import com.fasterxml.jackson.databind.introspect.TypeResolutionContext;
 import com.fasterxml.jackson.databind.introspect.VirtualAnnotatedMember;
 import com.fasterxml.jackson.databind.jsontype.TypeDeserializer;
@@ -80,6 +86,8 @@ import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 /**
  * A Jackson module that adds reflection-free bean serialization and deserialization for Micronaut.
@@ -106,6 +114,18 @@ public class BeanIntrospectionModule extends SimpleModule {
     public BeanIntrospectionModule() {
         setDeserializerModifier(new BeanIntrospectionDeserializerModifier());
         setSerializerModifier(new BeanIntrospectionSerializerModifier());
+    }
+
+    @Override
+    public void setupModule(SetupContext context) {
+        super.setupModule(context);
+
+        ObjectCodec owner = context.getOwner();
+        if (owner instanceof ObjectMapper) {
+            ObjectMapper mapper = (ObjectMapper) owner;
+            mapper.setConfig(mapper.getSerializationConfig().with(new BeanIntrospectionAccessorNamingStrategyProvider(mapper.getSerializationConfig().getAccessorNaming())));
+            mapper.setConfig(mapper.getDeserializationConfig().with(new BeanIntrospectionAccessorNamingStrategyProvider(mapper.getDeserializationConfig().getAccessorNaming())));
+        }
     }
 
     /**
@@ -371,7 +391,7 @@ public class BeanIntrospectionModule extends SimpleModule {
             }
 
             final Class<?> beanClass = beanDesc.getBeanClass();
-            final BeanIntrospection<Object> introspection = (BeanIntrospection<Object>) BeanIntrospector.SHARED.findIntrospection(beanClass).orElse(null);
+            final BeanIntrospection<Object> introspection = findIntrospection(beanClass);
             if (introspection == null) {
                 return builder;
             } else {
@@ -440,7 +460,7 @@ public class BeanIntrospectionModule extends SimpleModule {
                     @Override
                     public SettableBeanProperty[] getFromObjectArguments(DeserializationConfig config) {
 
-                        SettableBeanProperty[] existing = defaultInstantiator.getFromObjectArguments(config);
+                        SettableBeanProperty[] existing = ignoreReflectiveProperties ? null : defaultInstantiator.getFromObjectArguments(config);
                         if (props == null) {
                             props = new SettableBeanProperty[constructorArguments.length];
                             for (int i = 0; i < constructorArguments.length; i++) {
@@ -1008,6 +1028,52 @@ public class BeanIntrospectionModule extends SimpleModule {
         public Object setAndReturn(Object instance, Object value) {
             beanProperty.set(instance, value);
             return null;
+        }
+    }
+
+    private class BeanIntrospectionAccessorNamingStrategyProvider extends AccessorNamingStrategy.Provider {
+        private final AccessorNamingStrategy.Provider delegate;
+
+        BeanIntrospectionAccessorNamingStrategyProvider(AccessorNamingStrategy.Provider delegate) {
+            this.delegate = delegate;
+        }
+
+        @Override
+        public AccessorNamingStrategy forPOJO(MapperConfig<?> config, AnnotatedClass valueClass) {
+            return delegate.forPOJO(config, valueClass);
+        }
+
+        @Override
+        public AccessorNamingStrategy forBuilder(MapperConfig<?> config, AnnotatedClass builderClass, BeanDescription valueTypeDesc) {
+            return delegate.forBuilder(config, builderClass, valueTypeDesc);
+        }
+
+        @Override
+        public AccessorNamingStrategy forRecord(MapperConfig<?> config, AnnotatedClass recordClass) {
+            BeanIntrospection<Object> introspection = findIntrospection(recordClass.getRawType());
+            if (introspection != null) {
+                return new DefaultAccessorNamingStrategy(config, recordClass, null, "get", "is", null) {
+                    final Set<String> names = introspection.getBeanProperties().stream().map(BeanProperty::getName).collect(Collectors.toSet());
+
+                    @Override
+                    public String findNameForRegularGetter(AnnotatedMethod am, String name) {
+                        if (names.contains(name)) {
+                            return name;
+                        }
+                        return super.findNameForRegularGetter(am, name);
+                    }
+                };
+            } else {
+                try {
+                    return delegate.forRecord(config, recordClass);
+                } catch (IllegalArgumentException e) {
+                    if (e.getMessage().startsWith("Failed to access RecordComponents of type")) {
+                        throw new RuntimeException("Failed to construct AccessorNamingStrategy for record. This can happen when running in native-image. Either make this type @Introspected, or mark it for @ReflectiveAccess.", e);
+                    } else {
+                        throw e;
+                    }
+                }
+            }
         }
     }
 }

--- a/jackson-databind/src/test/groovy/io/micronaut/jackson/modules/BeanIntrospectionModuleRecordSpec.groovy
+++ b/jackson-databind/src/test/groovy/io/micronaut/jackson/modules/BeanIntrospectionModuleRecordSpec.groovy
@@ -1,0 +1,49 @@
+package io.micronaut.jackson.modules
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.micronaut.annotation.processing.test.AbstractTypeElementSpec
+import io.micronaut.context.ApplicationContext
+import io.micronaut.context.annotation.Replaces
+import io.micronaut.context.annotation.Requires
+import io.micronaut.core.beans.BeanIntrospection
+import jakarta.inject.Singleton
+import spock.lang.IgnoreIf
+
+@IgnoreIf({ !jvm.isJava14Compatible() })
+class BeanIntrospectionModuleRecordSpec extends AbstractTypeElementSpec {
+    def 'test record support'() {
+        given:
+        BeanIntrospection introspection = buildBeanIntrospection('test.Test', '''
+package test;
+import io.micronaut.core.annotation.Introspected;
+
+@Introspected
+record Test(String foo, String bar) {
+}
+''')
+        def ctx = ApplicationContext.run(['spec.name': 'BeanIntrospectionModuleRecordSpec'])
+        ctx.getBean(StaticBeanIntrospectionModule).introspectionMap[introspection.beanType] = introspection
+        ctx.getBean(BeanIntrospectionModule).ignoreReflectiveProperties = ignoreReflectiveProperties
+        def mapper = ctx.getBean(ObjectMapper)
+
+        when:
+        def value = mapper.readValue('{"foo":"1","bar":"2"}', introspection.beanType)
+        then:
+        value.foo == '1'
+        value.bar == '2'
+
+        where:
+        ignoreReflectiveProperties << [true, false]
+    }
+
+    @Singleton
+    @Replaces(BeanIntrospectionModule)
+    @Requires(property = "spec.name", value = 'BeanIntrospectionModuleRecordSpec')
+    static class StaticBeanIntrospectionModule extends BeanIntrospectionModule {
+        Map<Class, BeanIntrospection> introspectionMap = [:]
+        @Override
+        protected BeanIntrospection<Object> findIntrospection(Class<?> beanClass) {
+            return introspectionMap.get(beanClass)
+        }
+    }
+}

